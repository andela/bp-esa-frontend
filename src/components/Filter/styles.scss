.filter {
  position: relative;
  display: inline-block;
  font-size: 14px;
  margin-left: 25px;
  text-transform: capitalize;
  .filter-title {
    padding: 8px 20px 8px 20px;
    border: 1px solid grey;
    text-align: center;
    margin: 5px;
    color: #4d4d4d;
    cursor: pointer; 
    &:hover {
      color: black;
    }
    i {
      margin-left: 10px;
      font-size: 20px;
      transition: all 0.3s;
      &.rotate-180 {
        transform: rotate(180deg);
      }
    }
  }

  .filter-options {
    visibility: hidden;
    display: block;
    position: absolute;
    width: 120%;
    border-radius: 3px;
    background-color: white;
    box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.2);
    transition: all 0.3s;    
    li {
      border-bottom: 1px solid #d9d9d9;
      padding: 5px;
      input[type=checkbox] {
        margin-right: 10px;
      }
      .react-date-picker {
        width: 100%;
      }
    }
    li:last-child {
      border-bottom: none;
    }
    &.filter-options-isvisible {
      visibility: visible;
    }
  }
}