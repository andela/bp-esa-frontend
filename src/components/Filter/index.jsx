import React, { PureComponent } from 'react';
import classNames from 'classnames';
import PropTypes from 'proptypes';
import DatePicker from './DatePicker';
import './styles.scss';
import ReactDOM from 'react-dom';

class Filter extends PureComponent {
  constructor() {
    super();
    this.state = {
      filterOptionsIsVisible: false,
      selectedFilters: [],
      filterSet: '',
    };
  }

  componentDidMount() {
    const { filterSet } = this.props;
    this.setState({
      filterSet,
    });
  }

  /**
   * Toggle the display of the options available in the dropdown
   */
  toggleVisibility = (e) => {
    const { filterOptionsIsVisible } = this.state;

    // The DOM's UL element holding the list of options
    const filterOptionsEl = ReactDOM.findDOMNode(this).getElementsByClassName('filter-options')[0];

    // Handler for clicking anywhere else on the page
    const clickAway = (ev) => {
      if (ev.target === filterOptionsEl || filterOptionsEl.contains(ev.target)) return;
      if (e !== ev) {
        this.setState({ filterOptionsIsVisible: false });
        document.removeEventListener('click', clickAway);
      }
    };

    if (!filterOptionsIsVisible) {
      this.setState({ filterOptionsIsVisible: true });
      document.addEventListener('click', clickAway);
    }
  }

  selectCheckBoxFilter = (event) => {
    const { selectedFilters, filterSet } = this.state;
    const { handleFilterChange } = this.props;
    const filter = event.target.value;
    if (selectedFilters.includes(filter)) {
      const selectedFiltersToMutate = [...selectedFilters];
      selectedFiltersToMutate.splice(selectedFilters.indexOf(filter), 1);
      this.setState({
        selectedFilters: selectedFiltersToMutate,
      });
      handleFilterChange(filter, filterSet, 'remove_filter');
    } else {
      this.setState({
        selectedFilters: [...selectedFilters, event.target.value],
      });
      handleFilterChange(filter, filterSet, 'add_filter');
    }
  }

  selectDateFilter = (date, datePickerID) => {
    const { handleFilterChange } = this.props;
    const { filterSet } = this.state;
    if (datePickerID === `${1}`) {
      handleFilterChange(date, filterSet, 'set_from_date');
    } else if (datePickerID === `${2}`) {
      handleFilterChange(date, filterSet, 'set_to_date');
    }
  }

  selectFilterOption(option) {
    switch (option.type) {
      case 'checkbox':
        return this.renderCheckBoxOption(option);
      case 'date':
        return this.renderDateOption(option);
      default:
        return null;
    }
  }

  renderCheckBoxOption(option) {
    return (
      <li key={option.id}>
        <input type="checkbox" value={option.value} onChange={this.selectCheckBoxFilter} />
        <span>{option.text}</span>
      </li>
    );
  }

  renderDateOption(option) {
    return (
      <li key={option.id}>
        <DatePicker
          id={`${option.id}`}
          handleChange={this.selectDateFilter}
          placeholderText={option.text}
        />
      </li>
    );
  }


  renderFilterOptions() {
    const { options } = this.props;
    const { filterOptionsIsVisible } = this.state;
    return (
      <ul
        className={classNames('filter-options', { 'filter-options-isvisible': filterOptionsIsVisible })}
      >
        {options.map(option => (this.selectFilterOption(option)))}
      </ul>
    );
  }

  render() {
    const { title } = this.props;
    const { filterOptionsIsVisible } = this.state;
    return (
      <div className="filter">
        <div className="filter-title" onClick={this.toggleVisibility}>
          <span className="title">{title}</span>
          <i
            className={classNames('fa fa-angle-down', { 'rotate-180': filterOptionsIsVisible })}
          />
        </div>
        {this.renderFilterOptions()}
      </div>
    );
  }
}

Filter.propTypes = {
  filterSet: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  options: PropTypes.array.isRequired,
  handleFilterChange: PropTypes.func.isRequired,
};

export default Filter;
