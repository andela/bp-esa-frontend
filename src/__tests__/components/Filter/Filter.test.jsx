import React from 'react';
import Filter from '../../../components/Filter/index';

describe('Filter Component', () => {
  
  let props; 
  let component;
  let componentInstance;

  beforeEach(() => {
    props = {
      filterSet: 'automationStatus',
      title: 'Automation Status',
      options: [],
      handleFilterChange: jest.fn(),
    };
    component = (() => shallow(<Filter {...props} />))();
    componentInstance = component.instance();
  });

  it('should render as expected', () => {
    expect(component).toMatchSnapshot();
  });

  describe('toggleVisibility method', () => {
    it('should be initially "FALSE"', () => {
      expect(component.state('filterOptionsIsVisible')).toBeFalsy();
    });
  
    it('should change the filterOptionsIsVisible in the state when called', () => {
      const newprops = {
        ...props,
        options: [{ type: null }],
      };
      const component = mount(<Filter {...newprops} />);
      component.find('.filter-title').simulate('click');
      expect(component.state('filterOptionsIsVisible')).toBeTruthy();
    });
  });

  describe('selectCheckBoxFilter method', () => {
    it(`should update the selectedFilters in the state by adding
    the selected filter and also call handleFilterChange
    `, () => {
      const event = { target: { value: 'failed_automations' } };
      expect(component.state('selectedFilters')).toEqual([]);
      props.handleFilterChange.mockReset();
      componentInstance.selectCheckBoxFilter(event);
      expect(component.state('selectedFilters')).toEqual(['failed_automations']);
      expect(props.handleFilterChange).toHaveBeenCalledTimes(1);
    });

    it(`should update the selectedFilters in the state by removing
    the selected filter and also call handleFilterChange`, () => {
      const event = { target: { value: 'failed_automations' } };
      component.setState({ selectedFilters: ['failed_automations'] });
      props.handleFilterChange.mockReset();
      componentInstance.selectCheckBoxFilter(event);
      expect(component.state('selectedFilters')).toEqual([]);
      expect(props.handleFilterChange).toHaveBeenCalledTimes(1);
    });

    it('should check for default selectedFilters', () => {
      const newprops = {
        ...props,
        options: [{ type: null }],
      };
      const component = mount(<Filter {...newprops} />);
      component.find('.filter-title').simulate('click');
      expect(component.find('.filter-options-isvisible').length).toBe(1);
    });
  });

  describe('selectDateFilter method', () => {
    it(`should call handleFilterChange with set_from_date argument
    if the datePickerID is 1`, () => {
      const date = '1/02/2018';
      const datePickerID = '1';
      props.handleFilterChange.mockReset();
      componentInstance.selectDateFilter(date, datePickerID);
      expect(props.handleFilterChange).toHaveBeenCalledWith(date, props.filterSet, 'set_from_date');
    });

    it(`should call handleFilterChange with set_to_date argument
    if the datePickerID is 2`, () => {
      const date = '1/02/2018';
      const datePickerID = '2';
      props.handleFilterChange.mockReset();
      componentInstance.selectDateFilter(date, datePickerID);
      expect(props.handleFilterChange).toHaveBeenCalledWith(date, props.filterSet, 'set_to_date');
    });
  });
});
