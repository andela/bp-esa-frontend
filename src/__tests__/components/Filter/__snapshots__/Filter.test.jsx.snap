// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filter /> should render as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Filter
    filterSet="automationStatus"
    handleFilterChange={[MockFunction]}
    options={Array []}
    title="Automation Status"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="filter-title"
          onClick={[Function]}
        >
          <span
            className="title"
          >
            Automation Status
          </span>
          <i
            className="fa fa-angle-down"
          />
        </div>,
        <ul
          className="filter-options"
        />,
      ],
      "className": "filter",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="title"
            >
              Automation Status
            </span>,
            <i
              className="fa fa-angle-down"
            />,
          ],
          "className": "filter-title",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Automation Status",
              "className": "title",
            },
            "ref": null,
            "rendered": "Automation Status",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "fa fa-angle-down",
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "filter-options",
        },
        "ref": null,
        "rendered": Array [],
        "type": "ul",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="filter-title"
            onClick={[Function]}
          >
            <span
              className="title"
            >
              Automation Status
            </span>
            <i
              className="fa fa-angle-down"
            />
          </div>,
          <ul
            className="filter-options"
          />,
        ],
        "className": "filter",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="title"
              >
                Automation Status
              </span>,
              <i
                className="fa fa-angle-down"
              />,
            ],
            "className": "filter-title",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Automation Status",
                "className": "title",
              },
              "ref": null,
              "rendered": "Automation Status",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "fa fa-angle-down",
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "filter-options",
          },
          "ref": null,
          "rendered": Array [],
          "type": "ul",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Filter Component should render as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Filter
    filterSet="automationStatus"
    handleFilterChange={[MockFunction]}
    options={Array []}
    title="Automation Status"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="filter-title"
          onClick={[Function]}
        >
          <span
            className="title"
          >
            Automation Status
          </span>
          <i
            className="fa fa-angle-down"
          />
        </div>,
        <ul
          className="filter-options"
        />,
      ],
      "className": "filter",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="title"
            >
              Automation Status
            </span>,
            <i
              className="fa fa-angle-down"
            />,
          ],
          "className": "filter-title",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Automation Status",
              "className": "title",
            },
            "ref": null,
            "rendered": "Automation Status",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "fa fa-angle-down",
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "filter-options",
        },
        "ref": null,
        "rendered": Array [],
        "type": "ul",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="filter-title"
            onClick={[Function]}
          >
            <span
              className="title"
            >
              Automation Status
            </span>
            <i
              className="fa fa-angle-down"
            />
          </div>,
          <ul
            className="filter-options"
          />,
        ],
        "className": "filter",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="title"
              >
                Automation Status
              </span>,
              <i
                className="fa fa-angle-down"
              />,
            ],
            "className": "filter-title",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Automation Status",
                "className": "title",
              },
              "ref": null,
              "rendered": "Automation Status",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "fa fa-angle-down",
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "filter-options",
          },
          "ref": null,
          "rendered": Array [],
          "type": "ul",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
